src/shared/procedural_map/grid_map.rs

```diff
@@ -0,0 +1,171 @@
+        _width: TILE_WIDTH,
+        _height: TILE_HEIGHT,
+    });
+}
+
+pub fn setup_grid_and_spawn_tiles(
+    mut commands: Commands,
+    mut generate_map_event_reader: EventReader<GenerateMapEvent>,
+    tile_config: Res<TileConfig>,
+) {
+    if generate_map_event_reader.is_empty() {
+        return;
+    }
+
+    for generate_map_event in generate_map_event_reader.read() {
+        // Create a CaveConfig instance
+        let cave_config = CaveConfig {
+            map_width: DEFAULT_WIDTH,
+            map_height: DEFAULT_HEIGHT,
+            random_seed: generate_map_event.0,
+            empty_space_percentage: 40.0,
+            edge_thickness: 1,
+            max_dig_attempts: 10000,
+        };
+
+        // Generate the cave map
+        let mut new_cave_map = create_cave_map(
+            GridMap::new(DEFAULT_WIDTH as u32, DEFAULT_HEIGHT as u32),
+            cave_config,
+        );
+
+        // Insert the new GridMap resource
+        commands.insert_resource(new_cave_map.clone());
+
+        // Spawn tiles in filled cells
+        let mut tile_pool_index = 0;
+        let mut new_tiles = Vec::new();
+        let mut filled_count = 0;
+        let mut empty_count = 0;
+
+        for (i, state) in new_cave_map.states.iter().enumerate() {
+            let position = match state {
+                CellState::Empty => {
+                    empty_count += 1;
+                    continue;
+                }
+                CellState::Filled => {
+                    filled_count += 1;
+                    Vec2::new(
+                        (i as u32 % new_cave_map.width) as f32 * new_cave_map.width as f32,
+                        (i as u32 / new_cave_map.width) as f32 * new_cave_map.width as f32,
+                    )
+                }
+            };
+
+            match new_cave_map.tile_pool.get(tile_pool_index) {
+                // Reuse tile pool.
+                Some(entity) => {
+                    commands
+                        .entity(*entity)
+                        .insert(Transform::from_xyz(position.x, position.y, 0.0));
+                }
+                // If not enough in tile pool, spawn batch.
+                None => {
+                    new_tiles.push(
+                        commands
+                            .spawn(ColorMesh2dBundle {
+                                mesh: tile_config.mesh.clone(),
+                                material: tile_config.material.clone(),
+                                transform: Transform::from_xyz(position.x, position.y, 0.0),
+                                ..default()
+                            })
+                            .id(),
+                    );
+                }
+            };
+            tile_pool_index += 1;
+        }
+
+        new_cave_map.tile_pool.append(&mut new_tiles);
+
+        // Debug output for filled and empty cells
+        println!("Filled cells: {}", filled_count);
+        println!("Empty cells: {}", empty_count);
```

Comment:
```suggestion
        info!("Filled cells: {}", filled_count);
        info!("Empty cells: {}", empty_count);
```
Or remove it if it's not useful anymore.